### ISSUE TITLE
Odd variance for handler functions

### ISSUE THREAD

**Issue by matejvasek** [cloudevents/sdk-go#583]
Handler functions for `client.StartReceiver(ctx, handler)` could look like
```
	// This call is blocking.
	// Valid fn signatures are:
	// * func()
	// * func() error
	// * func(context.Context)
	// * func(context.Context) protocol.Result
	// * func(event.Event)
	// * func(event.Event) protocol.Result
	// * func(context.Context, event.Event)
	// * func(context.Context, event.Event) protocol.Result
	// * func(event.Event) *event.Event
	// * func(event.Event) (*event.Event, protocol.Result)
	// * func(context.Context, event.Event) *event.Event
	// * func(context.Context, event.Event) (*event.Event, protocol.Result)
```

In addition both inputs and outputs are covariant.
I think that input parameters should be either contravariant or invariant.

Handler function now can look like:
```go
func Handle(ctx *MyCtxImpl, event cloudevents.Event) *MyResult { return nil; }
```
where `MyCtxImpl` is a structure implementing `context.Context` and `MyResult` is a structure implementing `protocol.Result`.
The `StartReceiver` functions happily accept that function. However it will crash since the instance passed to the handler probably won't be `MyCtxImpl`. From type system point of view it would make more sense to have handlers like:
```go
func Handle(ctx interface{}, event cloudevents.Event) *MyResult { return nil; }
func Handle(ctx context.Context, event event.EventReader) *MyResult { return nil; }
```
however such a handlers won't be accepted by `StartReceiver` -- it will immediately panic.

**Comment by matejvasek** [comment]
Well, specifically for `ctx` covariance make sense since programmer knows exactly what type he or she passed to it.

**Comment by n3wscott** [comment]
This is making up for the lack of generics in golang. If used correctly (one of the signatures in the comment) I think it is pretty powerful. 

Is it a case or more documentation needed? or the pattern is just too complicated? It could use more reflection testing internally to catch the cases you mention.

**Comment by slinkydeveloper** [comment]
@matejvasek can you provide a use case for that? Usually context implementations are private and you extract elements from it using `ctx.Value()`

**Comment by matejvasek** [comment]
@n3wscott 

IMO this should be changed:
```diff
diff --git a/v2/client/receiver.go b/v2/client/receiver.go
index e1d1544..9b40cb7 100644
--- a/v2/client/receiver.go
+++ b/v2/client/receiver.go
@@ -113,15 +113,15 @@ func (r *receiverFn) validateInParamSignature(fnType reflect.Type) error {
        switch fnType.NumIn() {
        case 2:
                // has to be (context.Context, event.Event)
-               if !fnType.In(1).ConvertibleTo(eventType) {
+               if !eventType.ConvertibleTo(fnType.In(1)) { // Had to be inverted
                        return fmt.Errorf("%s; cannot convert parameter 2 from %s to event.Event", inParamUsage, fnType.In(1))
                } else {
                        r.hasEventIn = true
                }
                fallthrough
        case 1:
-               if !fnType.In(0).ConvertibleTo(contextType) {
-                       if !fnType.In(0).ConvertibleTo(eventType) {
+               if !contextType.ConvertibleTo(fnType.In(0)) { // Had to be inverted
+                       if !eventType.ConvertibleTo(fnType.In(0)) { // Had to be inverted
                                return fmt.Errorf("%s; cannot convert parameter 1 from %s to context.Context or event.Event", inParamUsage, fnType.In(0))
                        } else if r.hasEventIn {
                                return fmt.Errorf("%s; duplicate parameter of type event.Event", inParamUsage)
@@ -150,15 +150,15 @@ func (r *receiverFn) validateOutParamSignature(fnType reflect.Type) error {
        switch fnType.NumOut() {
        case 2:
                // has to be (*event.Event, transport.Result)
-               if !fnType.Out(1).ConvertibleTo(resultType) {
+               if !fnType.Out(1).ConvertibleTo(resultType) { // This is OK
                        return fmt.Errorf("%s; cannot convert parameter 2 from %s to event.Response", outParamUsage, fnType.Out(1))
                } else {
                        r.hasResultOut = true
                }
                fallthrough
        case 1:
-               if !fnType.Out(0).ConvertibleTo(resultType) {
-                       if !fnType.Out(0).ConvertibleTo(eventPtrType) {
+               if !fnType.Out(0).ConvertibleTo(resultType) { // This is OK
+                       if !fnType.Out(0).ConvertibleTo(eventPtrType) { // This is OK
                                return fmt.Errorf("%s; cannot convert parameter 1 from %s to *event.Event or transport.Result", outParamUsage, fnType.Out(0))
                        } else {
                                r.hasEventOut = true
```

I think `ConvertibleTo()` is used correctly for `Out()`, but I believe they should be inverted for `In()`.

**Comment by matejvasek** [comment]
```go
func Handle(ctx interface{}, event cloudevents.Event) *MyResult { return nil; }
```
This is really bad example from my side, I don't think it's any useful, but I think it's more type correct.

**Comment by n3wscott** [comment]
There are a ton of tests for this, would you have time adding your expected working examples here: https://github.com/cloudevents/sdk-go/blob/master/v2/client/receiver_test.go#L14

And then making the switch you suggest in https://github.com/cloudevents/sdk-go/issues/583#issuecomment-692941112 ?

Thanks for looking into it!! I can get to it as some point if you can't.

**Comment by matejvasek** [comment]
resolved by #587
